### ISSUE TITLE
`new` constraint not working as expected(?)

### ISSUE THREAD

**Issue by disarray2077** [beefytech/Beef#1204]
What I was expecting with this code is:
- `String` is a class, so it would obviously use the Allocate 1
- `A` is a constructable value type, so it should use Allocate 2
- `int` isn't a constructable value type, so it should use Allocate 3
This is what happens when you try to construct an `int` like a normal struct:
![image](https://user-images.githubusercontent.com/86157825/139596469-071ca1f8-0d71-4bd7-809b-5805ed5ded73.png)

But what ends up happening is that all value types use the method constrained only with `struct` (Allocate 3), this results in the `A` constructor not being called.
I guess the issue here is either `int` not being constructible like a normal struct and/or Allocate 3 being used instead of Allocate 2, or maybe just the lack of a constraint that would only allow struct initialization (and not pointer creation like `new` does).

```
	static
	{
		static Test<String> aaaaaaaaaaaa = .(); // Uses 'Allocate' 1
		static Test<A> aaaaaaaaaaa = .(); // Uses 'Allocate' 3
		static Test<int> aaaaaaaaaa = .(); // Uses 'Allocate' 3
	}

	struct A
	{
		public int a;
		public bool b;

		public this()
		{
			a = 10;
			b = false;
		}
	}

	struct Test<T>
		where T : var
	{
		// 1
		mixin Allocate() where T : class 
		{
			new T()
		}
		
		// 2
		mixin Allocate() where T : new, struct
		{
			T()
		}

		// 3
		mixin Allocate() where T : struct
		{
			default(T)
		}

		T a = Allocate!();
	}
```
Tested with: https://github.com/beefytech/Beef/commit/7a8592268c6c907acc2a8442f80f04bfbc3e073d

**Extra**
If you remove `where T : var` and add the `new` constraint to the Allocate 1, it gives you the `Must add 'where T : new'" error anyway! _(This looks like one of the issues in https://github.com/beefytech/Beef/issues/1194 but I already verified that all issues in that issue are fixed in my current build)_

Repro code:
```
	struct Test<T>
	{
		mixin Allocate() where T : new, class // << See method declaration at line 84:9 (this line)
		{
			new T()
		}
		
		mixin Allocate() where T : new, struct
		{
			T()
		}

		mixin Allocate() where T : struct
		{
			default(T)
		}

		T a = Allocate!(); // Must add 'where T : new' constraint
	}
```

**Comment by bfiete** [comment]
On the `extra` - an error is expected here. The `T` in the `Test<T>` definition doesn't match any of those mixins constraints. The first mixin is selected by default and then you get a specific constraint mismatch error. The error is saying that the actual `T` declaration requires the constraint (ie: `struct Test<T> where T : new`)

Technically I suppose the error could be more descriptive, perhaps, with extern constraints.
`Must add 'where T : new' constraint to `Test<T>` in order to use type as parameter 'T' for 'Test<T>.Allocate!()'`

With a `var` constraint we allow for and silently ignore that inability to select a method in the definition.

**Comment by bfiete** [comment]
Fixed at 4188f607c7c79d7256dfd3887e2425a71b31414b
There was a method selection issue, and also primitive types have default ctors (ie: `var f = float();` is equivalent to `float f = 0.0f;`). Doesn't make that much sense in isolation, but it's useful for generics.

**Comment by disarray2077** [comment]
> On the `extra` - an error is expected here. The `T` in the `Test<T>` definition doesn't match any of those mixins constraints. The first mixin is selected by default and then you get a specific constraint mismatch error. The error is saying that the actual `T` declaration requires the constraint (ie: `struct Test<T> where T : new`)
> 
> Technically I suppose the error could be more descriptive, perhaps, with extern constraints. `Must add 'where T : new' constraint to 'Test<T>' in order to use type as parameter 'T' for 'Test<T>.Allocate!()'`
> 
> With a `var` constraint we allow for and silently ignore that inability to select a method in the definition.

I see, this makes sense, I thought it should compile because this code compiles:
```
	struct Test<T>
	{
		mixin Allocate() where T : new, class
		{
			new T() // ERROR: Must add where T : new
		}

		mixin Allocate() where T : new, struct
		{
			T() // ERROR: Must add where T : new, struct
		}

		mixin Allocate()
		{
			default(T)
		}

		T a = Allocate!();
	}
```

But I just noticed that what was happening is that the compiler wasn't selecting the `mixin Allocate() where T : new, class` because there was another mixin without constraints
